[MLIR][TORCH] Support 2-d index for aten.index_put_impl op

vivekkhandelwal1 opened the pull request:
Signed-Off By: Vivek Khandelwa<vivek@nod-labs.com>

ramiro050 commented on lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:
```diff
@@ -275,14 +277,63 @@ class ConvertAten_IndexPutImplOp
+            indexTensorType.getDtype());
+        Value torchCstZero = rewriter.create<Torch::ConstantIntOp>(
+            loc, rewriter.getI64IntegerAttr(0));
+        indexTensor = rewriter.create<AtenSqueezeDimOp>(
```
I think we can just lower to `AtenFlattenUsingIntsOp`, and this way the implementation works for all 2-d index inputs, as well as higher ranked tensors

ramiro050 commented on lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:
```diff
@@ -275,14 +277,66 @@ class ConvertAten_IndexPutImplOp
+    if (inputRank != valuesRank) {
+      auto valueTensorType = op.getValues().getType().cast<BaseTensorType>();
+      SmallVector<int64_t> valueTensorShape(valueTensorType.getSizes());
+      if (valuesRank == (inputRank + 1) && valueTensorShape[0] == 1) {
```
This is also very restrictive. I would restrict things to the following case:

```python
if index_tensor.rank() > 1:
    if values.rank() != index_tensor.rank() + input.rank() - 1:
        return matchFailure("...")
    collapsed_index = torch.flatten(index_tensor)
    collapsed_values = torch.flatten(values, end_dim=index_tensor.rank() - 1)
```

This should handle any `index_tensor` and `values` tensor as long as `values.rank() == index_tensor.rank() + input.rank() - 1` is satisfied

ramiro050 commented on python/torch_mlir_e2e_test/test_suite/index_put.py:
```diff
@@ -818,3 +818,28 @@ def forward(self, input, index, value):
+    @export
+    @annotate_args([
+        None,
+        ([4, 7], torch.float32, True),
+        ([1, 3], torch.int64, True),
+        ([1, 3, 7], torch.float32, True),
```
If you use the less restrictive approach, you should no longer need to have static shapes here

vivekkhandelwal1 commented:
Hi @ramiro050, I need to support the following case for `aten._index_put_impl` op:
```
>>> a = torch.randn((1, 1, 2, 2))
>>> a
tensor([[[[ 1.4184, -1.8768],
          [-1.9334,  0.6857]]]])
>>> updates = torch.randn((1,1,4,4))
>>> updates
tensor([[[[ 1.2037,  1.7812, -0.2039,  0.3809],
          [-0.6564, -0.4894,  0.8012,  0.9277],
          [-1.2519, -1.1837,  1.0080, -0.3857],
          [ 1.8249, -2.4080,  0.0611,  0.7709]]]])
>>> index1 = torch.randint(high=2, size=(4, 1))
>>> index1
tensor([[0],
        [1],
        [1],
        [0]])
>>> index2 = torch.randint(high=2, size=(4,))
>>> index2
tensor([0, 0, 1, 1])
>>> result = torch.ops.aten._index_put_impl(a, (None, None, index1, index2), updates, accumulate=True, unsafe=False)
>>> result
tensor([[[[ 3.8203, -0.8679],
          [-5.5148,  3.0369]]]])

```
Do you think with the `tm_tensor.scatter` op, it would be possible? Since the `tm_tensor.scatter` op has many constraints of its own.

ramiro050 commented:
This should be doable with the `tm_tensor.scatter` op, although it is going to be a very tricky lowering. You need to transform the index tensors into the single index tensor that `tm_tensor.scatter` expects. An example of how this might be done is:

```python
import torch

input = torch.rand((2, 3, 4, 5))
index_dim2 = torch.randint(4, (6, 1))
index_dim3 = torch.randint(5, (7,))

# For the case of `torch.index_put(input, (None, None, index1, index2), ...)`
# TODO: compute this using numpy rules
# See: https://numpy.org/doc/stable/user/basics.indexing.html#combining-advanced-and-basic-indexing
final_shape = (input.size(0), input.size(1), 6, 7)

index_dim0_broadcast = torch.arange(0, input.size(0))\
                            .reshape(input.size(0), 1, 1, 1)\
                            .broadcast_to(final_shape)\
                            .flatten()
index_dim1_broadcast = torch.arange(0, input.size(1))\
                            .reshape(input.size(1), 1, 1)\
                            .broadcast_to(final_shape)\
                            .flatten()
index_dim2_broadcast = index_dim2.broadcast_to(final_shape).flatten()
index_dim3_broadcast = index_dim3.broadcast_to(final_shape).flatten()

indices = torch.stack([index_dim0_broadcast, index_dim1_broadcast,
                       index_dim2_broadcast, index_dim3_broadcast], dim=0).t()

print(indices.shape)
print(index_dim2)
print(index_dim3)
for index in indices:
    print(index)
```

This will create a tensor of size `(2 * 3 * 6 * 7, 4)`. You can think of it as `2 * 3 * 6 * 7 = 252` indices. Each index is `4` elements long because our input tensor is rank-4. @Shukla-Gaurav is a good person to consult on this since he had to do almost the same thing for `torch.scatter` (https://github.com/llvm/torch-mlir/pull/1499). I think we have support for most of the torch ops used in my example above, and in the `TTorchToMTensor` pass you're allowed to create more PyTorch ops, so hopefully this approach is not too bad.

Once you have the index tensor, you just need to flatten the `updates` tensor, and pass that to `tm_tensor.scatter`.

To determine the `final_shape` value in the example above, you need to be familiar with the indexing rules of numpy. [This link](https://numpy.org/doc/stable/user/basics.indexing.html#combining-advanced-and-basic-indexing) has information on it. In particular, this section:

> Let x.shape be (10, 20, 30, 40, 50) and suppose ind_1 and ind_2 can be broadcast to the shape (2, 3, 4). Then x[:, ind_1, ind_2] has shape (10, 2, 3, 4, 40, 50) because the (20, 30)-shaped subspace from X has been replaced with the (2, 3, 4) subspace from the indices. However, x[:, ind_1, :, ind_2] has shape (2, 3, 4, 10, 30, 50) because there is no unambiguous place to drop in the indexing subspace, thus it is tacked-on to the beginning.

I would reduce the problem to only handling the case where the index tensors are along consecutive dimensions, since this is the case you need for your example.

vivekkhandelwal1 commented:
Thanks @ramiro050 for such a detailed response!

ramiro050 commented on lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:
```diff
@@ -265,42 +268,256 @@ class ConvertAten_IndexPutImplOp
+          indexTensorType.getDtype());
+      indexTensor = rewriter.create<AtenUnsqueezeOp>(
+          loc, expandedIndexTensorType, indexTensor, torchCstOne);
+    } else if (indicesList.size() > 1 && indicesList.size() == 4) {
```
This is too restrictive. I don't think it would require much more to have full support  of 2 index tensors. We should have a helper function that computes the broadcasted size of the index tensors (not sure if this already exists in torch-mlir) to then compute `final_shape`. Then it should be a matter of turning every `None_i` in the index list into a `torch.arange(0, input.size(i)).reshape(1, ..., input.size(i), ..., 1, 1).flatten()`.

vivekkhandelwal1 replied:
Hi @ramiro050, could you please review this patch now?

ramiro050 replied:
Sure! I'm a bit busy right now, but I will for sure review it tomorrow.

vivekkhandelwal1 replied:
Thanks!

ramiro050 commented on lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:
```diff
@@ -410,42 +413,306 @@ class ConvertAten_IndexPutImplOp
+    BaseTensorType indexTensorType;
+    bool needToFlattenValuesTensor = false;
+    SmallVector<int64_t> flattenedValuesTensorSize;
+    if (indicesList.size() == 1) {
```
Why do we need to have a special case for 1 index tensor? I would've expected the logic for 2 index tensors to work for 1. Moreover, using the code for 2 index tensors would allow removing the restrictions on `indexTensorRank`

ramiro050 commented on lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:
```diff
@@ -410,42 +413,306 @@ class ConvertAten_IndexPutImplOp
+      indexTensor = rewriter.create<AtenUnsqueezeOp>(
+          loc, expandedIndexTensorType, indexTensor, torchCstOne);
+    } else if (indicesList.size() > 1 &&
+               (int64_t)indicesList.size() == inputRank) {
```
This limitation is not needed. You just need to pad the `indicesList` with `None`s

ramiro050 commented on lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:
```diff
@@ -410,42 +413,306 @@ class ConvertAten_IndexPutImplOp
+      // input tensor. Also, the index list should have atmost two non none
+      // index tensors along the consecutive dimension.
+
+      // This implementation assumes input and index tensors to be statically
+      // shaped.
```
We should be able to determine the broadcasted shape using dynamic dimensions. We just need to create runtime asserts for dynamic dimensions and create `aten.size.int` ops to get the size of the dynamic dimension.

ramiro050 commented on lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:
```diff
@@ -410,42 +413,306 @@ class ConvertAten_IndexPutImplOp
+      //                        dim=0).t()
+
+      Type indicesDtype =
+          indicesList[2].getType().cast<BaseTensorType>().getDtype();
```
Why 2? I don't think `indicesList` is guaranteed to have size 3 or larger

ramiro050 commented on lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:
```diff
@@ -410,42 +413,306 @@ class ConvertAten_IndexPutImplOp
+      flattenedValuesTensorLastDim = inputLastDim;
+    } else {
+      return rewriter.notifyMatchFailure(op,
+                                         "Indices list must not be empty and "
+                                         "its size could be either 1 or 4.");
```
This is not right

ramiro050 reviewed (changes requested):
A few more comments

vivekkhandelwal1 commented:
Hi @ramiro050, I have made the changes as per your comments. Right now, the changes work only when the shapes are static. I am working on adding support when the shapes are dynamic. That might require changes in some other op' lowering as well since the implementation creates view-like ops. But, there's one more issue, if I remove the code for handling cases with only 1 index tensor differently, the implementation is failing for the cases like this:
```
class IndexPutImpl2DIndexModule(torch.nn.Module):

    def __init__(self):
        super().__init__()

    @export
    @annotate_args([
        None,
        ([4, 7], torch.float32, True),
        ([2, 3], torch.int64, True),
        ([2, 3, 7], torch.float32, True),
    ])
    def forward(self, input, index, value):
        return torch.ops.aten._index_put_impl_(input, (index, ),
                                               value,
                                               accumulate=True,
                                               unsafe=False)


@register_test_case(
    module_factory=lambda: IndexPutImpl2DIndexModule())
def IndexPutImpl2DIndexModule_basic(module, tu: TestUtils):
    module.forward(tu.rand(4, 7), tu.randint(2, 3, high=3), tu.rand(2, 3, 7))
```
And, I need this case to be supported. Do you think the current generalized implementation can somehow handle this case as well?

vivekkhandelwal1 commented:
> Hi @ramiro050, I have made the changes as per your comments. Right now, the changes work only when the shapes are static. I am working on adding support when the shapes are dynamic. That might require changes in some other op' lowering as well since the implementation creates view-like ops. But, there's one more issue, if I remove the code for handling cases with only 1 index tensor differently, the implementation is failing for the cases like this:
> 
> ```
> class IndexPutImpl2DIndexModule(torch.nn.Module):
> 
>     def __init__(self):
>         super().__init__()
> 
>     @export
>     @annotate_args([
>         None,
>         ([4, 7], torch.float32, True),
>         ([2, 3], torch.int64, True),
>         ([2, 3, 7], torch.float32, True),
>     ])
>     def forward(self, input, index, value):
>         return torch.ops.aten._index_put_impl_(input, (index, ),
>                                                value,
>                                                accumulate=True,
>                                                unsafe=False)
> 
> 
> @register_test_case(
>     module_factory=lambda: IndexPutImpl2DIndexModule())
> def IndexPutImpl2DIndexModule_basic(module, tu: TestUtils):
>     module.forward(tu.rand(4, 7), tu.randint(2, 3, high=3), tu.rand(2, 3, 7))
> ```
> 
> And, I need this case to be supported. Do you think the current generalized implementation can somehow handle this case as well?

Hi @ramiro050, I have addressed all the issues. But in the process of making it generalized it has started unsupporting this particular case, as I have mentioned in the above comment. Do you have any idea regarding how this can be handled? Since we don't want to handle this as a separate case.

ramiro050 commented:
> Do you have any idea regarding how this can be handled? Since we don't want to handle this as a separate case.

Let me take a look.

vivekkhandelwal1 commented on lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:
```diff
@@ -409,43 +412,327 @@ class ConvertAten_IndexPutImplOp
+
+      // Check for broadcast shape size equal to number of non-none index
+      // tensors.
+      if (indexBroadcastShape.size() != nonNoneIndexTensorDim.size()) {
```
Thanks for the response! This implementation covers all the required cases.

ramiro050 commented on lib/Conversion/Utils/Utils.cpp:
```diff
@@ -347,6 +347,100 @@ Value toPositiveValidDim(ConversionPatternRewriter &rewriter, Location loc,
+  if (rankA == 0 && rankB == 0) {
+    // Do nothing.
+    return success();
+  } else if (rankA && rankB) {
```
I think the code in this section works regardless of whether some of the inputs are zero rank or not. No need for the `if-else` statements

ramiro050 commented on lib/Conversion/TorchToArith/TorchToArith.cpp:
```diff
@@ -305,38 +305,38 @@ class ConvertAtenAnyOrAllBoolOp : public OpConversionPattern<OpTy> {
                   ConversionPatternRewriter &rewriter) const override {
-
+    Location loc = op.getLoc();
+    SmallVector<bool> inputListBool;
```
nit: unused

ramiro050 commented on lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:
```diff
@@ -360,6 +360,222 @@ class ConvertAtenBincountOp : public OpConversionPattern<AtenBincountOp> {
 };
 } // namespace
 
+static SmallVector<SmallVector<int64_t>>
```
Nit: can you add a small documentation string to this function? Without the context of the Python script, it's not very obvious what this function does

ramiro050 commented on lib/Conversion/Utils/Utils.cpp:
```diff
@@ -347,6 +347,100 @@ Value toPositiveValidDim(ConversionPatternRewriter &rewriter, Location loc,
+
+// Checks whether the `shapeA` and `shapeB` are broadcast compatible or not. If
+// yes, then computes the final broadcast shape.
+LogicalResult computeBroadcastShape(ConversionPatternRewriter &rewriter,
```
LogicalResult return type no longer needed. All paths lead to `success`

ramiro050 commented on lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:
```diff
@@ -401,51 +623,211 @@ class ConvertAten_IndexPutImplOp
-      return rewriter.notifyMatchFailure(op, "Index tensor must not be None.");
+    if (sizeOptionalIndicesList == 0) {
+      return rewriter.notifyMatchFailure(op, "Indices list must not be empty.");
+    } else {
```
The previous `if` returns early. No need to have `else` branch (decreases indentation)

ramiro050 commented on lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:
```diff
@@ -401,51 +623,211 @@ class ConvertAten_IndexPutImplOp
+    } else {
+      for (unsigned i = 0; i < optionalIndicesList.size(); i++) {
+        if (!optionalIndicesList[i].getType().isa<Torch::NoneType>()) {
+          // indicesList.push_back(optionalIndicesList[i]);
```
not needed

ramiro050 reviewed (approved):
Great work on this! I just have a few small comments

vivekkhandelwal1 commented:
> Great work on this! I just have a few small comments

Thanks, @ramiro050, for your continuous feedback and help, without which it would have been hard to get this working.
